### PR TITLE
pkg/manifests: rotate grpc key material gracefully

### PR DESCRIPTION
**Pull request by s-urbaniak**
* [ ] I added CHANGELOG entry for this change.
* [x] No user facing changes, so no entry in CHANGELOG was needed.

This is the initial implementation for a graceful rotation based on the simplified scheme. Please take a look. It still misses a small e2e test which enforces rotation.

/cc @openshift/openshift-team-monitoring

### CODE REVIEW

**Comment by s-urbaniak** [comment]
the failure
```
    --- FAIL: TestUserWorkloadMonitoring/assert_user_workload_metrics (10.16s)
        client.go:270: unexpected status code response, want 200, got 403 ("\n<!DOCTYPE html>\n<html lang=\"en\" charset=\"utf-8\">\n<head>\n  <title>Log In</title>\n  <meta name=\"viewport\" content=\"width=device-width, initial-scale=1, maximum-scale=1, user-scalable=no\">\n  <style>\n    @font-face {\n      font-family: \"Open Sans\";\n      src: url(data:application/x-font-woff;charset=utf-8;base64,d09GRgABAAAAAFeoABMAAAAAlkQAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAABGRlRNAAABqAAAABwAAAAcavCZq0dERUYAAAHEAAAAHQAAAB4AJwD1R1BPUwAAAeQAAASjAAAJni1yF0JHU1VCAAAGiAAAAIEAAACooGKInk9TLzIAAAcMAAAAYAAAAGCh3ZrDY21hcAAAB2wAAAGcAAACAv1rbL5jdnQgAAAJCAAAADIAAAA8K3MG4GZwZ20AAAk8AAAE+gAACZGLC3pBZ2FzcAAADjgAAAAIAAAACAAAABBnbHlmAAAOQAAAQG0AAHBIDuDVH2hlYWQAAE6wAAAANAAAADYHgk2EaGhlYQAATuQAAAAgAAAAJA37BfVobXR4AABPBAAAAjgAAAO8MaBM1GxvY2EAAFE8AAAB1QAAAeB9N5qybWF4cAAAUxQAAAAgAAAAIAMhAjxuYW1lAABTNAAAAeQAAARWRvKTBXBvc3QAAFUYAAAB+AAAAvgEbWOAcHJlcAAAVxAAAACQAAAAkPNEIux3ZWJmAABXoAAAAAYAAAAGxDVUvgAAAAEAAAAA0MoNVwAAAADJQhegAAAAANDkdLN42mNgZGBg4AFiMSBmYmAEwndAzALmMQAADdgBHQAAAHjarZZNTJRHGMf/uyzuFm2RtmnTj2hjKKE0tikxAb...")
    user_workload_monitoring_test.go:65: scenario "assert user workload metrics" failed
```
seems to be unrelated, as it is not using grpc key material, but let's see if this failure is consistent.

File: pkg/manifests/tls.go [hunk 1]
```diff
@@ -36,94 +39,145 @@ func needsNewCert(notBefore, notAfter time.Time, now func() time.Time) bool {
...
+			fmt.Println("##### certs expired")
 		}
-		s.Namespace = f.namespace
-		s.Data = make(map[string][]byte)
-		s.Annotations = make(map[string]string)
 	}
 
-	crt, crtPresent := s.Data["ca.crt"]
-	key, keyPresent := s.Data["ca.key"]
+	if _, ok := s.Annotations["monitoring.openshift.io/grpc-forced-rotate"]; ok {
```
Comment (paulfantom) [inline pkg/manifests/tls.go:84]:
```suggestion
	if _, ok := s.Annotations["monitoring.openshift.io/grpc-forced-cert-rotate"]; ok {
```
Let's be a little bit more explicit what it is about

Comment (s-urbaniak) [reply pkg/manifests/tls.go:84]: it's not just the cert, it is the complete key material, including server/client certs and keys, hence i suggest `grpc-tls-forced-rotate` to gather the whole context.

File: pkg/manifests/tls.go [hunk 2]
```diff
@@ -36,94 +39,145 @@ func needsNewCert(notBefore, notAfter time.Time, now func() time.Time) bool {
...
+	if crtPresent && keyPresent {
 		var err error
-		s, err = f.NewSecret(MustAssetReader(ClusterMonitoringGrpcTLSSecret))
+		curCA, err = crypto.GetCAFromBytes(curCABytes, curCAKeyBytes)
 		if err != nil {
-			return nil, err
+			klog.Warningf("creating a new CA due to error reading CA: %v", err)
+		} else if needsNewCert(curCA.Config.Certs[0].NotBefore, curCA.Config.Certs[0].NotAfter, time.Now) {
+			rotate = true
+			fmt.Println("##### certs expired")
```
Comment (s-urbaniak) [inline pkg/manifests/tls.go:80]: whoopsie

Comment (lilic) [reply pkg/manifests/tls.go:80]: Happens to me as well often, its WIP. And no curse words included so all good ;)

File: test/e2e/user_workload_monitoring_test.go [hunk 3]
```diff
@@ -59,6 +59,9 @@ func TestUserWorkloadMonitoring(t *testing.T) {
 		{"assert user workload rules", assertUserWorkloadRules},
 		{"assert tenancy model is enforced", assertTenancyForMetrics},
 		{"assert prometheus and alertmanager is not deployed in user namespace", assertPrometheusAlertmanagerInUserNamespace},
+		{"assert grpc tls rotation", assertGRPCTLSRotation},
+		{"assert user workload metrics", assertUserWorkloadMetrics},
+		{"assert user workload rules", assertUserWorkloadRules},
```
Comment (s-urbaniak) [inline test/e2e/user_workload_monitoring_test.go:64]: cc @openshift/openshift-team-monitoring i am lacking a bit of creativity here tbh :thinking: do you have other ideas how we could test this, other than a) enforcing rotation and b) asserting if everything "still" works?

**Comment by s-urbaniak** [comment]
Another failure in `TestMultinamespacePrometheusRule`, investigating.

File: test/e2e/user_workload_monitoring_test.go (hunk 3 above)
Comment (lilic) [reply test/e2e/user_workload_monitoring_test.go:64]: Should we be adding this to this existing test? I would prefer a separate test case here. TestGRPCTLSUserWorkloadMonitoring maybe?

**Comment by s-urbaniak** [comment]
build is green, but I would like to execute a couple more runs.

File: test/e2e/user_workload_monitoring_test.go (hunk 3 above)
Comment (s-urbaniak) [reply test/e2e/user_workload_monitoring_test.go:64]: Sounds good, that makes things a bit more complex though, as tests run in parallel and I'd like to execute rotation in the context of user workload monitoring too.

I'd need to refactor then the e2e test suite here so we test two (potentially concurrently running) user workload monitoring tests one dedicated for rotation, the other for asserting metrics/alerts. Do you think the added complexity is worth it? If yes, I'd go ahead.

Comment (lilic) [reply test/e2e/user_workload_monitoring_test.go:64]: I just feel like this test is growing, that is my main concern and we are testing too much in one test which would make it not clear what is failing. Its not a blocker for me, happy to have a follow up PR in 4.7 or after feature freeze to refactor this as well, wdyt?

Comment (s-urbaniak) [reply test/e2e/user_workload_monitoring_test.go:64]: Yeah, I agree, I feel on the edge here, the list is still well readable but I agree we should start thinking about a potential refactor. I submitted https://issues.redhat.com/browse/MON-1170 to track this and would suggest to tackle it as a follow-up.

Comment (lilic) [reply test/e2e/user_workload_monitoring_test.go:64]: Thanks! That sounds good!

File: pkg/manifests/tls.go [hunk 4]
```diff
@@ -36,94 +39,144 @@ func needsNewCert(notBefore, notAfter time.Time, now func() time.Time) bool {
...
+		rotate                    = !crtPresent || !keyPresent
+	)
+
+	if crtPresent && keyPresent {
 		var err error
-		s, err = f.NewSecret(MustAssetReader(ClusterMonitoringGrpcTLSSecret))
+		curCA, err = crypto.GetCAFromBytes(curCABytes, curCAKeyBytes)
 		if err != nil {
-			return nil, err
+			klog.Warningf("creating a new CA due to error reading CA: %v", err)
```
[the remark this answers is not in the document]
Comment (s-urbaniak) [reply pkg/manifests/tls.go:77]: Yes, this is a quirky edge case. It is hard as we are indeed flying blind what went wrong here. For some reason reading the existing CA failed (maybe someone modified it manually, or any other corruption happened)?

We have two options:
a) bail out here and crash-loop and wait for the user to resolve the corrupted CA (not good)
b) what I implemented here: log out the issue and simply re-issue a new CA.

wdyt?

File: pkg/manifests/tls.go [hunk 5]
```diff
@@ -36,94 +39,144 @@ func needsNewCert(notBefore, notAfter time.Time, now func() time.Time) bool {
...
-	key, keyPresent := s.Data["ca.key"]
+	if _, ok := s.Annotations["monitoring.openshift.io/grpc-tls-forced-rotate"]; ok {
+		rotate = true
+		delete(s.Annotations, "monitoring.openshift.io/grpc-tls-forced-rotate")
+	}
 
-	if crtPresent && keyPresent {
-		ca, err := crypto.GetCAFromBytes(crt, key)
+	if !rotate {
+		return nil
```
[the remark this answers is not in the document]
Comment (s-urbaniak) [reply pkg/manifests/tls.go:89]: rotate is initialized already with `rotate = !crtPresent || !keyPresent`, but you're right, I missed setting it to true if reading failed :+1:  

Let me add a unit test for this case too.

[the remark this answers is not in the document]
Comment (s-urbaniak) [reply pkg/manifests/tls.go:89]: hmm, i do have a unit test for this edge case, let me recheck :thinking:

File: pkg/manifests/tls.go [hunk 6]
```diff
@@ -36,94 +39,144 @@ func needsNewCert(notBefore, notAfter time.Time, now func() time.Time) bool {
...
 		s.Data["prometheus-server.crt"] = crt
 		s.Data["prometheus-server.key"] = key
 	}
 
-	return s, nil
+	return nil
+}
+
+// CreateCertificate creates a new certificate and returns it in x509.Certificate form.
+func CreateCertificate(template, parent *x509.Certificate, pub, priv interface{}) (*x509.Certificate, error) {
```
[the remark this answers is not in the document]
Comment (s-urbaniak) [reply pkg/manifests/tls.go:172]: oh no reason, let me set it to private.

File: pkg/manifests/tls.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (s-urbaniak) [reply pkg/manifests/tls.go:89]: ha, great catch, the unit test is wrong and hence also the logic, fixing :tada:

[the remark this answers is not in the document]
Comment (s-urbaniak) [reply pkg/manifests/tls.go:89]: fixed

**Comment by s-urbaniak** [comment]
@openshift/openshift-team-monitoring ptal

**Comment by s-urbaniak** [comment]
> One question -> do we trigger the controller resync on this file?
yes as per https://github.com/openshift/cluster-monitoring-operator/blob/5e32ac396e5225901088a696ceebdf4fc3aedc40/pkg/operator/operator.go#L245

File: pkg/manifests/tls.go [hunk 7]
```diff
@@ -36,94 +39,150 @@ func needsNewCert(notBefore, notAfter time.Time, now func() time.Time) bool {
...
 
 	{
-		cfg, err := ca.MakeServerCert(
+		cfg, err := newCA.MakeServerCert(
 			sets.NewString("prometheus-grpc"),
 			crypto.DefaultCertificateLifetimeInDays,
 		)
 		if err != nil {
-			return nil, errors.Wrap(err, "error making prometheus-k8s server certificate")
+			return errors.Wrap(err, "error making prometheus-k8s server certificate")
```
Comment (lilic) [inline pkg/manifests/tls.go:163]: Nit: What does prometheus-k8s mean to users? 🤔 e.g. the one we deploy in openshift-cluster-monitoring, correct? To users might be helpful to include that?

File: pkg/manifests/tls_test.go [hunk 8]
```diff
@@ -88,23 +93,74 @@ func TestNeedsNewCert(t *testing.T) {
...
+				if !reflect.DeepEqual(spre.Data, spost.Data) {
+					return fmt.Errorf("expected certificate data to be equal, but it isn't, pre-rotation key material %v, post-rotation key material %v", spre, spost)
+				}
+				return nil
+			},
+		},
+		{
+			name: "force rotation",
+			setup: func(s *v1.Secret) {
+				s.Annotations["monitoring.openshift.io/grpc-tls-forced-rotate"] = "true"
```
Comment (lilic) [inline pkg/manifests/tls_test.go:113]: Could we add an Annotation with`blah/foo-bar`? Would it make sense to have that testcase?

File: pkg/tasks/clustermonitoringoperator.go [hunk 9]
```diff
@@ -72,19 +72,25 @@ func (t *ClusterMonitoringOperatorTask) Run() error {
...
+	s, err := t.factory.GRPCSecret()
+	if err != nil {
+		return errors.Wrap(err, "error initializing Cluster Monitoring Operator GRPC TLS secret")
 	}
 
-	if err != nil {
+	loaded, err := t.client.GetSecret(s.Namespace, s.Name)
+	switch {
+	case apierrors.IsNotFound(err):
+		// No secret was found, proceed with the default empty secret from manifests.
```
Comment (lilic) [inline pkg/tasks/clustermonitoringoperator.go:83]: Should we add debug logs here besides comments?

File: pkg/tasks/clustermonitoringoperator.go [hunk 10]
```diff
@@ -72,19 +72,25 @@ func (t *ClusterMonitoringOperatorTask) Run() error {
...
+	case err == nil:
+		// Secret was found, use that.
```
Comment (lilic) [inline pkg/tasks/clustermonitoringoperator.go:85]: Same question as above, up to you if you think its needed as you were developing this.

File: pkg/manifests/tls.go [hunk 11]
```diff
@@ -36,94 +39,150 @@ func needsNewCert(notBefore, notAfter time.Time, now func() time.Time) bool {
 	return now().After(latestTime)
 }
 
-// This method creates a central secret containing GRPC TLS key material
-// if the given secret is nil
-// or updates it if the CA present in the given secret is about to expire.
+func (f *Factory) GRPCSecret() (*v1.Secret, error) {
```
Comment (lilic) [inline pkg/manifests/tls.go:42]: I wonder if we can move this to the manifests.go now, that is purely just manifests generation? Not urgent to address just as we have this pattern already.

File: pkg/manifests/tls.go [hunk 12]
```diff
@@ -36,94 +39,150 @@ func needsNewCert(notBefore, notAfter time.Time, now func() time.Time) bool {
...
+			rotate = true
 		}
-		s.Namespace = f.namespace
-		s.Data = make(map[string][]byte)
-		s.Annotations = make(map[string]string)
 	}
 
-	crt, crtPresent := s.Data["ca.crt"]
-	key, keyPresent := s.Data["ca.key"]
+	if _, ok := s.Annotations["monitoring.openshift.io/grpc-tls-forced-rotate"]; ok {
```
[the remark this answers is not in the document]
Comment (s-urbaniak) [reply pkg/manifests/tls.go:89]: Personally not a fan of this, as those literals are intrinsic constants already (we don't follow that idiom elsewhere here and in many golang projects too), especially given the fact that this is not a library used externally.

File: pkg/manifests/tls.go (hunk 7 above)
Comment (s-urbaniak) [reply pkg/manifests/tls.go:163]: You have a good point, we actually have only two certificate, one server cert (used in thanos ruler and prometheus) and one client cert. I will use these terms instead.

File: pkg/manifests/tls_test.go (hunk 8 above)
Comment (s-urbaniak) [reply pkg/manifests/tls_test.go:113]: The expected behavior would be a no-op, not sure about the benefit, but I can add a test case

File: pkg/tasks/clustermonitoringoperator.go (hunk 9 above)
Comment (s-urbaniak) [reply pkg/tasks/clustermonitoringoperator.go:83]: that's a great idea :+1:

File: pkg/manifests/tls.go (hunk 11 above)
Comment (s-urbaniak) [reply pkg/manifests/tls.go:42]: Yeah, i was back and forth on this one. It could be considered the _first_ step to untangle `manifests.go` which is too big already. I would leave it here and consequently cut `manifests.go` in separate files as a follow-up.

**Comment by s-urbaniak** [comment]
@lilic PTAL and unhold if you're fine with the proposed changes.

File: pkg/manifests/tls.go (hunk 11 above)
Comment (lilic) [reply pkg/manifests/tls.go:42]: But the test for it is in `pkg/manifests/manifests_test.go `, I would then either move this function there or move the test to tls_test.go?

Comment (s-urbaniak) [reply pkg/manifests/tls.go:42]: let's move the test to `tls_test.go` indeed and set a first cutting precedence :+1:
